haystack/components/evaluators/document_map.py

```diff
@@ -0,0 +1,89 @@
+from typing import Any, Dict, List
+
+from haystack import Document, component
+
+
+@component
+class DocumentMeanAveragePrecision:
+    """
+    Evaluator that calculates the mean average precision of the retrieved documents.
+    Each question can have multiple ground truth documents and multiple retrieved documents.
+
+    Usage example:
+    ```python
+    from haystack.components.evaluators import AnswerExactMatchEvaluator
+
+    evaluator = DocumentMeanAveragePrecision()
+    result = evaluator.run(
+        questions=[
+            "In what country is Normandy located?",
+            "When was the Latin version of the word Norman first recorded?",
+        ],
+        ground_truth_documents=[
+            [Document(content="France")],
+            [Document(content="9th century"), Document(content="9th")],
+        ],
+        retrieved_documents=[
+            [Document(content="France")],
+            [Document(content="9th century"), Document(content="10th century"), Document(content="9th")],
+        ],
+    )
+
+    print(result["individual_scores"])
+    # [1.0, 0.8333333333333333]
+    print(result["score"])
+    # 0.9166666666666666
+    ```
+    """
+
+    @component.output_types(score=float, individual_scores=List[float])
+    def run(
+        self,
+        questions: List[str],
```

Comment:
I'm wondering if we need to include questions as part of the input at all - we can simply use `len(retrieved_docs)` instead (and make it clear in the docstrings that each inner list corresponds to the output/GT pair for a single question). 

Same with the exact answer match, incidentally.

Reply:
We already discussed this didn't we? 

I remember we decided to keep the `questions` in any case to make it clear that the other data is related to the questions.

I do agree it's completely useless though.